[IsDeprecatedWeakRefSmartPointerException] Remove exception from WebRTCMonitor and LibWebRTCSocket

RupinMittal opened the pull request:
#### 9e6935faed8654391f8d9d3d4ff78cef98a91bac
<pre>
[IsDeprecatedWeakRefSmartPointerException] Remove exception from WebRTCMonitor and LibWebRTCSocket
<a href="https://bugs.webkit.org/show_bug.cgi?id=281115">https://bugs.webkit.org/show_bug.cgi?id=281115</a>
<a href="https://rdar.apple.com/137565108">rdar://137565108</a>

Reviewed by Chris Dumez and Ryosuke Niwa.

We remove IsDeprecatedWeakRefSmartPointerException by making WebRTCMonitor ref-counted.

In the case of LibWebRTCSocket, we cannot make it ref-counted because
LibWebRTCSocketFactory::createUdpSocket and LibWebRTCSocketFactory::createClientTcpSocket
return a raw pointer of LibWebRTCSocket&apos;s base class type rtc::AsyncPacketSocket, which
could be destroyed--completely ignoring the ref-counting of LibWebRTCSocket. So we remove
the exception by making LibWebRTCSocket able to be a CheckedPtr.

* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
(WebKit::LibWebRTCNetworkManager::getOrCreate):
(WebKit::LibWebRTCNetworkManager::signalUsedInterface):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.h:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCProvider.cpp:
(WebKit::LibWebRTCProvider::createPeerConnection):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocket.h:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocketFactory.cpp:
(WebKit::LibWebRTCSocketFactory::forSocketInGroup):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocketFactory.h:
* Source/WebKit/WebProcess/Network/webrtc/WebRTCMonitor.cpp:
(WebKit::WebRTCMonitor::networksChanged):
(WebKit::WebRTCMonitor::networkProcessCrashed):
* Source/WebKit/WebProcess/Network/webrtc/WebRTCMonitor.h:

Canonical link: <a href="https://commits.webkit.org/285004@main">https://commits.webkit.org/285004@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/37ed0ed43bc29f02f55a870ec6195e82aed232bf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71139 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; check-webkit-style") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50553 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23913 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75247 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22344 "Built successfully") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58353 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22164 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56247 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14718 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74205 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45947 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61320 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36683 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42620 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20685 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64512 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19142 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76967 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15374 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18326 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63972 "Passed tests") | 
| | [⏳ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61360 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63939 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/15880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12081 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5711 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92169 "Build is in progress. Recent messages:") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10914 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46353 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1132 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/92169 "Build is in progress. Recent messages:") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47424 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48707 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47166 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/WebProcess/Network/webrtc/WebRTCMonitor.h:
```diff
@@ -53,6 +44,10 @@ struct NetworksChangedData;
 public:
     virtual ~WebRTCMonitorObserver() = default;
+
+    virtual void ref() const = 0;
```
Use DECLARE_VIRTUAL_REFCOUNTED macro?

cdumez replied:
I still don't like those macros.

rniwa commented on Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.h:
```diff
@@ -32,16 +32,20 @@
+    static RefPtr<LibWebRTCNetworkManager> getOrCreate(WebCore::ScriptExecutionContextIdentifier);
     ~LibWebRTCNetworkManager();
 
+    void ref() const final { RefCounted::ref(); }
```
Use DEFINE_VIRTUAL_REFCOUNTED macro?

cdumez replied:
I like being explicit about it. I find the macros confusing.

cdumez commented on Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocketFactory.cpp:
```diff
@@ -102,7 +102,7 @@ rtc::AsyncPacketSocket* LibWebRTCSocketFactory::createClientTcpSocket(WebCore::S
     }
 
-    return socket.release();
+    return socket.leakRef();
```
Personally, I doubt this patch is correct (cc @youennf). Calling leakRef() here and passing it to libwebrtc as a `rtc::AsyncPacketSocket* ` likely means libwebrtc will call `delete` on the object when they don't need it anymore, completely ignoring our ref-counting?

RupinMittal replied:
Would the correct thing be to use `releaseNonNull`?

cdumez replied:
We're implementing an interface required by libwebrtc, which expects a `rtc::AsyncPacketSocket*` so you cannot return a smart pointer. They expect a raw pointer to the base class. libwebrtc also doesn't know that the object is ref-counted so I'm pretty sure they'll just end up calling `delete` when done.

I think this object (and the other one above) simply CANNOT be ref-counted since we don't control the way the object is used / destroyed.

RupinMittal replied:
Ok that does make sense. Hopefully this can be fixed by making LibWebRTCSocket able to be a CheckedPtr. We'll see if the bots are happy with this.

cdumez reviewed (changes requested):
r- because I don't think this is safe (@youennf can confirm).

cdumez commented on Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocket.h:
```diff
@@ -58,8 +49,9 @@ namespace WebKit {
 class LibWebRTCSocketFactory;
 
-class LibWebRTCSocket final : public rtc::AsyncPacketSocket, public CanMakeWeakPtr<LibWebRTCSocket>, public Identified<WebCore::LibWebRTCSocketIdentifier> {
+class LibWebRTCSocket final : public CanMakeCheckedPtr<LibWebRTCSocket>, public rtc::AsyncPacketSocket, public Identified<WebCore::LibWebRTCSocketIdentifier> {
```
We should leave `rtc::AsyncPacketSocket` as *first* base class, it is much safer. Otherwise, you could end up introducing crashes if existing code relies on reinterpret_cast<>() or conversion to and from `void*`.

cdumez replied:
If you're curious this may shade some light: https://www.sjbrown.co.uk/posts/always-use-static-cast/

cdumez reviewed (approved):
LGTM but please fix the base class order.

RupinMittal committed:
[IsDeprecatedWeakRefSmartPointerException] Remove exception from WebRTCMonitor and LibWebRTCSocket
https://bugs.webkit.org/show_bug.cgi?id=281115
rdar://137565108

Reviewed by Chris Dumez and Ryosuke Niwa.

We remove IsDeprecatedWeakRefSmartPointerException by making WebRTCMonitor ref-counted.

In the case of LibWebRTCSocket, we cannot make it ref-counted because
LibWebRTCSocketFactory::createUdpSocket and LibWebRTCSocketFactory::createClientTcpSocket
return a raw pointer of LibWebRTCSocket's base class type rtc::AsyncPacketSocket, which
could be destroyed--completely ignoring the ref-counting of LibWebRTCSocket. So we remove
the exception by making LibWebRTCSocket able to be a CheckedPtr.

* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.cpp:
(WebKit::LibWebRTCNetworkManager::getOrCreate):
(WebKit::LibWebRTCNetworkManager::signalUsedInterface):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCNetworkManager.h:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCProvider.cpp:
(WebKit::LibWebRTCProvider::createPeerConnection):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocket.h:
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocketFactory.cpp:
(WebKit::LibWebRTCSocketFactory::forSocketInGroup):
* Source/WebKit/WebProcess/Network/webrtc/LibWebRTCSocketFactory.h:
* Source/WebKit/WebProcess/Network/webrtc/WebRTCMonitor.cpp:
(WebKit::WebRTCMonitor::networksChanged):
(WebKit::WebRTCMonitor::networkProcessCrashed):
* Source/WebKit/WebProcess/Network/webrtc/WebRTCMonitor.h:

Canonical link: https://commits.webkit.org/285004@main
